### ISSUE TITLE
Make nifty handle InputEvents but not consume them

### ISSUE THREAD

**Issue by oberien** [nifty-gui/nifty-gui#274]
I'd like to use nifty in a HUD in front of an in lwjgl rendered game. The problem is, that nifty consumes all inputEvents. By calling Nifty.setIgnoreMouseEvents is doesn't handle these events anymore, as it returns false in the beginning of processMouseEvent. Is there a way to achieve this?

If not, this will be a quick sourcecode-edit implementing this feature. I didn't test it and therefore I don't know, if it actually works, but for me it seems like a solution. In case there is no other way, can someone open a pull request pls @3xp0n3nt?

Changes in nifty-core/src/main/java/de/lessvoid/nifty/Nifty.java

Insert this between line 140 and 141:

``` java
/*
 * when set to false Nifty won't consume MouseEvents after handling them.
*/
private boolean consumeMouseEvents;

/*
 * when set to false Nifty won't consume InputEvents after handling them.
*/
private boolean consumeInputEvents;

```

Between 1935 and 1936:

``` java
public void setConsumeMouseEvents(final boolean newValue) {
    consumeMouseEvents = newValue;
}

public boolean isConsumeMouseEvents() {
    return consumeMouseEvents;
}
public void setConsumeKeyboardEvents(final boolean newValue) {
    consumeKeyboardEvents = newValue;
}

public boolean isConsumeMouseEvents() {
    return consumeKeyboardEvents;
}

```

Replace line 1566 width:

``` java
return processed && consumeMouseEvents;
```

Replace line 1582 with:

``` java
return processed && consumeKeyboardEvents;
```

Regards,
BH16

**Comment by relu91** [comment]
I think this is covered in the manual, but only about Slick2d implementation ... read the section INPUT FORWARDING under INTEGRATION WITH OTHER SYSTEMS chapter. I think that your solution is a bit simple and maybe doesn't fit well for a general propose case. The right way to do could be working with InputSystems as it was done with Slick2d.

**Comment by ghost** [comment]
@Blackhole16 I just implemented this for the LibGDX backend. Check this out:

https://github.com/void256/nifty-gui/commit/c3e2591ca92080bff98ba993069b735ede889091

You're probably not using LibGDX, but is that the idea you're after?

**Comment by oberien** [comment]
I found the Error, it is working now.

@relu91 This InputSystem for whatever reason haven't be working. Therefore I implemented a manual input forwarding described in the 13th answer in this thread: http://slick.ninjacave.com/forum/viewtopic.php?t=3436 (Posted: Fri Jul 01, 2011 3:13 pm by arielsan).

@3xp0n3nt Exactly that is what I WAS looking for.

As I use the StateBasedGame's implemented methods (mouseMoved/pressed/released, keyPressed/Released) to create and forward a MouseEvent to nifty, Slick's input forwarding doesnt work anymore. As it seems the StateBasedGame catches all InputEvents and forwards them to the current State. Because I override them, there is no forwarding to the currentGameState anymore. So my solution is to call the super-method before forwarding the events to Nifty.

So this problem actually had nothing to do with Nifty's implementation of EventConsuming :)

Anyway thank's for the help and the ideas.
Kind Regards,
BH16
